src/spatial_data_format.rs

```diff
@@ -0,0 +1,515 @@
+                    })
+                    .collect()
+            })
+            .collect();
+        Self::attach_tags(&self.header.tags, interpolated_streams)
+    }
+
+    /// Extracts the streams from the file, and quantizes data points to
+    /// produce data points at the requested sample_rate.
+    fn streams_quantized(&self, sample_rate: u64) -> Vec<(GrapeTag, Vec<f32>)> {
+        debug_assert!(sample_rate < self.header.sample_rate);
+        let pts_per_sample = self.header.sample_rate as usize / sample_rate as usize;
+        let raw_streams = self.get_raw_streams();
+        let quantized_streams = raw_streams
+            .into_iter()
+            .map(|v| {
+                v.chunks(pts_per_sample)
+                    .map(|c| c.iter().sum::<f32>() / c.len() as f32)
+                    .collect()
+            })
+            .collect();
+        Self::attach_tags(&self.header.tags, quantized_streams)
+    }
+}
+
+/// This builder contains the data required
+#[derive(Debug, Clone)]
+pub struct GrapeFileBuilder {
+    sample_rate: u64,
+    streams: Vec<(GrapeTag, Vec<f32>)>,
+}
+
+impl Default for GrapeFileBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GrapeFileBuilder {
+    /// Instaintiate a builder with no streams and a default sample rate of
+    /// 1000 samples per second.
+    fn new() -> Self {
+        GrapeFileBuilder {
+            sample_rate: 1000,
+            streams: Vec::new(),
+        }
+    }
+
+    /// Sets the sample rate of the builder to the argument.
+    pub fn set_samplerate(self, sample_rate: u64) -> Self {
+        GrapeFileBuilder {
+            sample_rate,
+            ..self
+        }
+    }
+
+    /// Adds a tagged stream to the builder.
+    pub fn add_stream(mut self, stream: &[f32], tag: GrapeTag) -> Self {
+        let stream: Vec<f32> = stream.to_vec();
+        self.streams.push((tag, stream));
+        self
+    }
+
+    /// Removes all streams from the builder
+    pub fn clear_streams(mut self) -> Self {
+        self.streams.clear();
+        self
+    }
+
+    /// Builds a [GrapeFile] from the builder, truncating all streams to the
+    /// length of the shortest stream.
+    pub fn build_truncate(self) -> GrapeFile {
+        let tags: Vec<GrapeTag> = self
+            .streams
+            .iter()
+            .map(|(tag, _vec)| tag)
+            .cloned()
+            .collect();
+        let sample_vecs: Vec<Vec<f32>> = self.streams.into_iter().map(|(_tag, vec)| vec).collect();
+
+        let mut samples = Vec::new();
+
+        if !sample_vecs.is_empty() {
+            let shortest = sample_vecs.iter().map(|v| v.len()).min().unwrap();
+            samples.reserve_exact(shortest * sample_vecs.len() as usize);
+            for sample_idx in 0..shortest {
+                for stream_idx in 0..sample_vecs.len() {
+                    samples.push(sample_vecs[stream_idx as usize][sample_idx]);
+                }
+            }
```

Comment:
Could we simplify this to:
```suggestion
            samples = sample_vecs.iter().map(|v| v.truncate(shortest));
```

Reply:
I wasn't quite able to get that to work, how about
```suggestion
let shortest = sample_vecs.iter().map(|v| v.len()).min().unwrap_or(0);
let samples = sample_vecs.iter().flat_map(|v| &v[0..shortest]).cloned().collect::<Vec<_>>();
```

Reply:
Not quite, since we need to interleave all the samples from each vec. So `samples` should contain `sample_vecs[0][0]`, then `sample_vecs[1][0]`, etc... Essentially we need to iterate the streams faster than the sample indices.

Reply:
I realized this after making the comment, and vscode didn't remove the comment in time 😅 makes total sense tho!